### ISSUE TITLE
[RFC] Make coercers work with parametarization of roles

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#1285]
I was expecting the coercer to coerce the Rat to an Int. If that's not possible to do, the role should complain about it during parametarization.

```
$ perl6 -e 'my role Coercer[::T] { method c(T $v) { $v } }; say Coercer[Int()].new.c: 42.12'
Type check failed in binding to parameter '$v'; expected Int(Any) but got Rat (42.12)
  in method c at -e line 1
  in block <unit> at -e line 1

$ perl6 -v
This is Rakudo version 2017.11-8-g7939014 built on MoarVM version 2017.11
implementing Perl 6.c.
```

**Comment by zoffixznet** [comment]
Explanation for why it won't work: https://irclog.perlgeek.de/perl6/2017-11-30#i_15513907

> the role should complain

.oO( but what if someone wants to pass coercers as arguments... )

**Comment by jnthn** [comment]
What I maybe didn't communicate so well: there's no reason this can't be made to work at some point in the future. It's just not entirely trivial to get there.

**Comment by zoffixznet** [comment]
Ah, OK. Then I'll reopen as an RFC :)

**Comment by vrurg** [comment]
Closed with merge of #3891
